common/src/main/java/org/conscrypt/ConscryptEngine.java

```diff
@@ -1081,39 +1215,64 @@ private SSLEngineResult readPendingBytesFromBIO(ByteBuffer dst, int bytesConsume
      */
     private int readEncryptedData(final ByteBuffer dst, final int pending) throws SSLException {
         try {
-            int bioRead = 0;
+            int bytesRead = 0;
+            final int pos = dst.position();
```

Comment:
Moving this outside the loop seems wrong.  For example, if we do multiple reads, it looks like line 1227 will be wrong on the second one, since it'll update the position to pos (the position before any reads) plus bytesRead (the amount read by the second read).